### ISSUE TITLE
Swiping through card while having multiple cities prone to errors

### ISSUE THREAD

**Issue by emkay5771** [breezy-weather/breezy-weather#168]
### Steps to reproduce

1) have multiple cities in city list
2) try going through hourly forecast
3) if not precise, will switch cities by hotizontal scrolling

### Expected behavior

Shouldn't be too easy to change pages 

### Actual behavior

Switches very easily, without intending to

### Weather provider used

Accuweather

### Breezy Weather version

4.1.3-Beta

### Android version

13

### Device

Pixel 7

### Other details

Perhaps add a settings toggle to disable the swiping between cities?

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to **[latest version](https://github.com/breezy-weather/breezy-weather/releases/latest)**.

**Comment by papjul** [comment]
Can you reproduce the issue in Geometric Weather or in Breezy Weather 4.0.x?

I did notice the swipe was sensitive, however I wasn't sure as when trying to reproduce in Geometric Weather, it seemed identical to me.

I'm asking that because in v4.1.0, I removed a duplicated code from an old Android SDK regarding horizontal paging, and original developer didn't leave notes as to why he did that. It was about 2000 lines long and felt overkill and mostly identical to SDK. If I can identify what he changed exactly, I can probably just extend the existing `ViewPager` and change only the few functions that he rewrote.

Best would be to make a diff from Android SDK version and Geometric Weather code, but as we don't know which SDK version he implemented and since there were multiple classes across many files, it would take hours if not days.

Tagging @WangDaYeeeeee in case he can add us some inputs.

**Comment by emkay5771** [comment]
Having tried it, it does seem to be very similar to the behavior within Geometric Weather. 

**CORRECTION**: This was tested with **4.1.3**, NOT 1.3.4. Apologies for the mixup

**Comment by emkay5771** [comment]
One more note: this sensitivity is particularly noticable when trying to interact and scrub through the 120 min. precipitation forecast. (To try a current example, New York currently has rain)

**Comment by papjul** [comment]
I don't have any issue at all on the precipitation graph, it doesn't even tilt a bit, see:

https://github.com/breezy-weather/breezy-weather/assets/1449159/28df31b8-b834-4bd6-8a25-e2166f9077c5

Please provide a video from your side (check your Android tiles for screen recording).

**Comment by emkay5771** [comment]
> I don't have any issue at all on the precipitation graph, it doesn't even tilt a bit, see:
> 
>  TRIM_20230716_203805.mp4 
> Please provide a video from your side (check your Android tiles for screen recording).

Do you have multiple cities in your list?
See this video 

https://github.com/breezy-weather/breezy-weather/assets/52011243/f822cbe9-17e1-42b5-9a97-35e6bb3c26cc

**Comment by min7-i** [comment]
I think there are two scenarios:
1. You tap on the little chip and hold your finger (see hightlighted area in screenshot). In that case you can safely move your finger left and right without changing location pages.</br>![precipitation](https://github.com/breezy-weather/breezy-weather/assets/97251923/f1f43968-f3db-4110-8343-f7d6cb73a0ef)
2. You're tapping on any other area within the precipitation card and swipe. In that case the chip will jump to the tapped position and depending on the intesity of your swipe you'll change the location page.

**Comment by emkay5771** [comment]
> I think there are two scenarios:
> 
> 1. You tap on the little chip and hold your finger (see hightlighted area in screenshot). In that case you can safely move your finger left and right without changing location pages.![precipitation](https://user-images.githubusercontent.com/97251923/253822735-f1f43968-f3db-4110-8343-f7d6cb73a0ef.png)
> 2. You're tapping on any other area within the precipitation card and swipe. In that case the chip will jump to the tapped position and depending on the intesity of your swipe you'll change the location page.

Unfortunately doesn't seem to be the case... See this clip 

https://github.com/breezy-weather/breezy-weather/assets/52011243/7a3b5cb4-96fc-423e-9a3e-d6edad5bd34b

**Comment by papjul** [comment]
> I think there are two scenarios:
> 
> 1. You tap on the little chip and hold your finger (see hightlighted area in screenshot). In that case you can safely move your finger left and right without changing location pages.![precipitation](https://user-images.githubusercontent.com/97251923/253822735-f1f43968-f3db-4110-8343-f7d6cb73a0ef.png)
> 2. You're tapping on any other area within the precipitation card and swipe. In that case the chip will jump to the tapped position and depending on the intesity of your swipe you'll change the location page.

Both scenarios are safe on my side.

I do have multiple cities, I only added New York as you mentioned it, so I still have my other cities.

**Comment by papjul** [comment]
Why do we see a little tap on your videos? Do you have a touch device or is it a mouse emulation or something?

**Comment by emkay5771** [comment]
> Why do we see a little tap on your videos? Do you have a touch device or is it a mouse emulation or something?

Pixel screen recording allows you to show taps on the screen... Generally more helpful to see the interaction points.

**Comment by min7-i** [comment]
> Unfortunately doesn't seem to be the case... See this clip

In that clip you're tapping **below** the area that I marked on my screenshot. Try tapping within the area that is surrounded by the red rectangle on my screenshot. (The purple chip displaying the 'x inch/h' in your clip.)

Just to add, in general, I can recreate the behavior that is shown in your clip on my Pixel 6a. That's the behavior I meant to describe with the 2. scenario.

**Comment by papjul** [comment]
Is it me or we have 3 different scenarios here?

| Who | Swiping on area | Swiping anywhere in the graph |
| --- | --- | --- |
| @emkay5771 | ❌ | ❌ |
| @min7-i | ✅ | ❌ |
| @papjul  | ✅ | ✅ |

Thank you Android for making consistent experiences across devices!

I will just end up adding another preference… even though it should probably have been disabled since the beginning. Having two horizontal swipe events in parallel is not a good UX IMO

**Comment by emkay5771** [comment]
I found the touch zone which allows me to swipe the precipitation map. It seems to be a really small touch target though. Thanks!

**Comment by papjul** [comment]
OK, I now understand why it felt more prone to errors than Geometric Weather.

In Geometric Weather, horizontal swipe between cities INSIDE cards does not trigger.

Given you have different behavior than me, if you can try to swipe anywhere on a card on Geometric Weather and let me know what happens on your side ;)

**Comment by emkay5771** [comment]
> OK, I now understand why it felt more prone to errors than Geometric Weather.
> 
> In Geometric Weather, horizontal swipe between cities INSIDE cards does not trigger.
> 
> Given you have different behavior than me, if you can try to swipe anywhere on a card on Geometric Weather and let me know what happens on your side ;)

That seems to be the case 👍

**Comment by papjul** [comment]
Nevermind, I'm silly, I did the test with only one city, of course it wouldn’t switch

**Comment by MarkWieczorek** [comment]
I am having a very similar issue when there are multiple locations set.

I note that if I swipe at roughly a 45 degree angle anywhere on the screen (including within cards), that this initiates a swipe to the next location. This is bad design, as it is very easy to make a mistake. From a user perspective, you probably should not allow swiping between locations within cards at all.

* Second, given that this bug makes having multiple locations unusable, I would suggest adding a setting to turn off swiping between locations altogether.

**Comment by papjul** [comment]
Swiping between locations is a very popular feature among weather apps.

Even when I added a poll asking if we should disable swiping by default given this major issue, people were still mixed in the result.

To me, it is because the only issue here is that we have two horizontal swipe (trends + between locations). Other than that, it should be expected that when swiping, you go to next location.

The biggest mystery to me is still why people never complained in more than 7 years of Geometric Weather about this, but now it seems to be a problem.

I think I need to review a few other weather apps to check how they are doing it but from an initial look, I seem to see that swiping usually require more than 50 % of the screen to be swiped, while in Geometric Weather/Breezy Weather, just a few % of the screen is enough to trigger the swipe. Increasing swipe "length" could be a first step to make it less prone to errors.

**Comment by MarkWieczorek** [comment]
What would you think of only allowing swipe actions in the header, and not in the underlying cards?

For me the length of the swipe has no influence: swiping just even a mm changes locations. The only factor that seems to influence this is the angle. Any angle greater than about 30 degrees with respect to horizontal seems to change location.

**Comment by papjul** [comment]
> For me the length of the swipe has no influence: swiping just even a mm changes locations. 

Required distance to swipe to next location is 20 % of screen width in v4.5.4. Additionally, when swipe is nested (daily/hourly trends), there is a minimum swipe distance required before it starts triggering so it’s even more than 20%.

So your issue is different; please open a new issue with a screen recording (please include touch points in it) about your location being switched before this 20 % + nested safety net margin.

____

I increased required swipe distance to 50 % for next version (v4.6.0) and with the nested distance safety net, it makes it safe in daily/hourly trends.
